### PR TITLE
Support of block.timestamp and block.number in smart contracts.

### PR DESCRIPTION
**Pull request by evgenyzdanovich**
Support of block.timestamp and block.number in smart contracts.

### CODE REVIEW

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0008334491808 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0008334726272 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00797264 | 0.0028341140672 | 4 | 2 | 4156320 | 3986320 | 95.91 | 
| Borrow from lending pool | 0.00477928 | 0.0016989384544000001 | 5 | 10 | 2644640 | 2389640 | 90.36 | 
| Repay | 0.00013 | 4.62124e-05 | 5 | 10 | 290000 | 65000 | 22.41 | 
| Flashloan | 0.00473928 | 0.0016847192544 | 4 | 7 | 2634640 | 2369640 | 89.94 | 
| Withdraw | 7e-05 | 2.48829e-05 | 5 | 5 | 250000 | 35000 | 14.0 | 
| Liquidation | 0.00247464 | 0.0008796355344000001 | 5 | 11 | 1752320 | 1237320 | 70.61 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0008334726272 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02747064 | 0.0097647136944 | 4 | 7 | 14000320 | 13735320 | 98.11 | 
| NonfungiblePositionManager - Increase liquidity  | 8e-05 | 2.8436e-05 | 23 | 6 | 230000 | 40000 | 17.39 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0004969894783999999 | 18 | 1 | 829040 | 699040 | 84.32 | 
| NonfungiblePositionManager - Collect Fees | 2e-05 | 7.109e-06 | 21 | 1 | 80000 | 10000 | 12.5 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 7.109e-06 | 16 | 1 | 50000 | 10000 | 20.0 | 
| Direct swap | 2e-05 | 7.109e-06 | 18 | 1 | 180000 | 10000 | 5.56 | 
| Burn transaction | 0.00047936 | 0.000170388512 | 12 | 1 | 279680 | 239680 | 85.7 | 
| Collect transaction | 0.00234464 | 0.000833402288 | 13 | 1 | 1202320 | 1172320 | 97.5 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0008369567879999999 | 4 | 2 | 1432320 | 1177320 | 82.2 | 
| Swap DAI -> USDC | 2e-05 | 7.109e-06 | 22 | 1 | 190000 | 10000 | 5.26 | 
| Swap USDC -> DAI | 2e-05 | 7.1092000000000005e-06 | 22 | 1 | 190000 | 10000 | 5.26 | 
| Remove liquidity | 2e-05 | 7.109e-06 | 24 | 1 | 100000 | 10000 | 10.0 | 
| Add liquidity in 3 Tokens | 0.00241464 | 0.0008583079344 | 4 | 7 | 1492320 | 1207320 | 80.9 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 7.109e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 7.109e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 7.109e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Remove liquidity 3 pool | 3e-05 | 1.06638e-05 | 4 | 2 | 180000 | 15000 | 8.33 | 
| Add liquidity in metapool | 0.00241464 | 0.0008582596415999999 | 4 | 7 | 1522320 | 1207320 | 79.31 | 
| Swap SUSD -> LP metapool | 2e-05 | 7.1092000000000005e-06 | 23 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> SUSD metapool | 0.00015 | 5.331749999999999e-05 | 5 | 11 | 560000 | 75000 | 13.39 | 
| Swap DAI -> USDT metapool | 0.00013 | 4.6207199999999994e-05 | 4 | 7 | 330000 | 65000 | 19.7 | 
| Remove liquidity in Metapool | 2e-05 | 7.1088000000000004e-06 | 24 | 1 | 100000 | 10000 | 10.0 |

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   58045 |      0     |      0.151 |    0.00144309  |        0      |
| getSlot                           |   57950 |      0     |      0.148 |    0.000823762 |        0      |
| getBlocks                         |   23279 |      0     |      0.049 |    0.000605739 |        0      |
| getBlock                          |   15071 |      0     |      0.08  |    0.00131975  |        0.001  |
| getMultipleAccounts               |   13315 |      0     |      0.152 |    0.00187195  |        0.001  |
| getTransaction                    |    7960 |      0     |      0.201 |    0.0100005   |        0.004  |
| getBlockTime                      |    6154 |      0     |      0.047 |    0.000493825 |        0      |
| sendTransaction                   |    5156 |      0     |      0.391 |    0.00775698  |        0.003  |
| getLatestBlockhash                |    4968 |      0     |      0.148 |    0.00203663  |        0.001  |
| getBalance                        |    1819 |      0     |      0.101 |    0.00156789  |        0      |
| getSignatureStatuses              |    1112 |      0     |      0.1   |    0.00130126  |        0      |
| getVersion                        |     580 |      0     |      0.082 |    0.00145862  |        0      |
| isBlockhashValid                  |     538 |      0     |      0.023 |    0.000408922 |        0      |
| getSignaturesForAddress           |      61 |      0     |      0.003 |    0.00106557  |        0.001  |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000526316 |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.002 |    0.0006875   |        0.001  |
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>


Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.000831409344 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02747064 | 0.0097413636504 | 4 | 7 | 14000320 | 13735320 | 98.11 | 
| NonfungiblePositionManager - Increase liquidity  | 8e-05 | 2.8368000000000004e-05 | 23 | 6 | 230000 | 40000 | 17.39 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0004957731487999999 | 18 | 1 | 829040 | 699040 | 84.32 | 
| NonfungiblePositionManager - Collect Fees | 2e-05 | 7.092000000000001e-06 | 21 | 1 | 80000 | 10000 | 12.5 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 7.092000000000001e-06 | 16 | 1 | 50000 | 10000 | 20.0 | 
| Direct swap | 2e-05 | 7.0922000000000005e-06 | 18 | 1 | 180000 | 10000 | 5.56 | 
| Burn transaction | 0.00047936 | 0.0001699858496 | 12 | 1 | 279680 | 239680 | 85.7 | 
| Collect transaction | 0.00234464 | 0.0008314327903999999 | 13 | 1 | 1202320 | 1172320 | 97.5 |

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0008416319743999999 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0008416085279999999 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00797264 | 0.0028618588544 | 4 | 2 | 4156320 | 3986320 | 95.91 | 
| Borrow from lending pool | 0.00477928 | 0.001715522556 | 5 | 10 | 2644640 | 2389640 | 90.36 | 
| Repay | 0.00013 | 4.66635e-05 | 5 | 10 | 290000 | 65000 | 22.41 | 
| Flashloan | 0.00473928 | 0.0017012119488000002 | 4 | 7 | 2634640 | 2369640 | 89.94 | 
| Withdraw | 7e-05 | 2.5127899999999997e-05 | 5 | 5 | 250000 | 35000 | 14.0 | 
| Liquidation | 0.00247464 | 0.0008883215208000001 | 5 | 11 | 1752320 | 1237320 | 70.61 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0008416554208 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02747064 | 0.0098614103472 | 4 | 7 | 14000320 | 13735320 | 98.11 | 
| NonfungiblePositionManager - Increase liquidity  | 8e-05 | 2.8718400000000003e-05 | 23 | 6 | 230000 | 40000 | 17.39 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0005018827584 | 18 | 1 | 829040 | 699040 | 84.32 | 
| NonfungiblePositionManager - Collect Fees | 2e-05 | 7.179400000000001e-06 | 21 | 1 | 80000 | 10000 | 12.5 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 7.179600000000001e-06 | 16 | 1 | 50000 | 10000 | 20.0 | 
| Direct swap | 2e-05 | 7.179600000000001e-06 | 18 | 1 | 180000 | 10000 | 5.56 | 
| Burn transaction | 0.00047936 | 0.00017208065280000003 | 12 | 1 | 279680 | 239680 | 85.7 | 
| Collect transaction | 0.00234464 | 0.0008416788672 | 13 | 1 | 1202320 | 1172320 | 97.5 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0008452686672 | 4 | 2 | 1432320 | 1177320 | 82.2 | 
| Swap DAI -> USDC | 2e-05 | 7.179600000000001e-06 | 22 | 1 | 190000 | 10000 | 5.26 | 
| Swap USDC -> DAI | 2e-05 | 7.179600000000001e-06 | 22 | 1 | 190000 | 10000 | 5.26 | 
| Remove liquidity | 3e-05 | 1.07694e-05 | 4 | 2 | 100000 | 15000 | 15.0 | 
| Add liquidity in 3 Tokens | 0.00241464 | 0.0008668074672000001 | 4 | 7 | 1492320 | 1207320 | 80.9 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 7.179400000000001e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 7.179600000000001e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 7.179600000000001e-06 | 22 | 1 | 200000 | 10000 | 5.0 | 
| Remove liquidity 3 pool | 3e-05 | 1.07694e-05 | 4 | 2 | 180000 | 15000 | 8.33 | 
| Add liquidity in metapool | 0.00241464 | 0.0008668074672000001 | 4 | 7 | 1522320 | 1207320 | 79.31 | 
| Swap SUSD -> LP metapool | 2e-05 | 7.179600000000001e-06 | 23 | 1 | 200000 | 10000 | 5.0 | 
| Swap USDC -> SUSD metapool | 0.00015 | 5.3847e-05 | 5 | 11 | 560000 | 75000 | 13.39 | 
| Swap DAI -> USDT metapool | 0.00011 | 3.9487800000000005e-05 | 4 | 6 | 330000 | 55000 | 16.67 | 
| Remove liquidity in Metapool | 3e-05 | 1.07694e-05 | 4 | 2 | 100000 | 15000 | 15.0 |

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   59614 |      0     |      0.095 |    0.000789294 |         0     |
| getAccountInfo                    |   59272 |      0     |      0.094 |    0.00127922  |         0     |
| getBlocks                         |   24316 |      0     |      0.039 |    0.000649449 |         0     |
| getBlock                          |   16752 |      0     |      0.062 |    0.00141052  |         0.001 |
| getMultipleAccounts               |   14281 |      0     |      0.094 |    0.00161774  |         0.001 |
| getLatestBlockhash                |   11707 |      0     |      0.135 |    0.00113035  |         0     |
| getTransaction                    |    8184 |      0     |      0.115 |    0.00779044  |         0.003 |
| getBlockTime                      |    6257 |      0     |      0.075 |    0.000452773 |         0     |
| sendTransaction                   |    5312 |      0     |      0.114 |    0.00582869  |         0.003 |
| getBalance                        |    1852 |      0     |      0.038 |    0.00132343  |         0.001 |
| getSignatureStatuses              |    1174 |      0     |      0.086 |    0.00112777  |         0     |
| getVersion                        |     613 |      0     |      0.087 |    0.00147961  |         0     |
| isBlockhashValid                  |     569 |      0     |      0.014 |    0.000434095 |         0     |
| getSignaturesForAddress           |      63 |      0     |      0.005 |    0.00134921  |         0.001 |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000263158 |         0     |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00059375  |         0.001 |
| getTokenAccountBalance            |       4 |      0     |      0.001 |    0.00025     |         0     |
| getHealth                         |       3 |      0.002 |      0.006 |    0.00366667  |         0.003 |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

File: evm_loader/lib/src/commands/emulate.rs
Comment (anton-lisanin) [inline evm_loader/lib/src/commands/emulate.rs:1]: Entire block with `if storage.is_timestamp_used() {` need to be removed.

File: evm_loader/program/src/executor/state.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub type TouchedAccounts = TreeMap<Pubkey, u64>;
 /// Persistent part of `ExecutorState`.
 #[repr(C)]
 pub struct ExecutorStateData {
-    cache: RefCell<Cache>,
+    pub cache: Cache,
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: Please keep it as `RefCell`.
It is required for https://github.com/neonlabsorg/neon-evm/pull/553

File: evm_loader/program/src/instruction/transaction_step.rs [hunk 2]
```diff
@@ -117,15 +117,27 @@ fn allocate_or_reinit_state(
...
 
-        // Dealloc objects that were potentially alloced in previous iterations before the reset.
+        // Dealloc evm that was potentially alloced in previous iterations before the reset.
         if storage.is_evm_alloced() {
             storage.dealloc_evm::<EvmBackend, NoopEventListener>();
         }
+
+        // Dealloc executor state that was potentially alloced in previous iterations before the reset.
+        // Also, copy the cache for use into the new ExecutorStateData.
+        let mut cache = None;
```
Comment (anton-lisanin) [inline evm_loader/program/src/instruction/transaction_step.rs]: The cache may contain other data in the future. But here we only need to restore the timestamp and the block.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

File: evm_loader/lib/src/commands/emulate.rs
Comment (evgenyzdanovich) [reply evm_loader/lib/src/commands/emulate.rs:1]: done

File: evm_loader/program/src/executor/state.rs (hunk 1 above)
Comment (evgenyzdanovich) [reply evm_loader/program/src/executor/state.rs]: done. Moved out `block_timestamp` and `block_number` into a separate struct - `BlockParams`.

`Cache` will stay as it is right now (`Refcell`'ed)

File: evm_loader/program/src/instruction/transaction_step.rs (hunk 2 above)
Comment (evgenyzdanovich) [reply evm_loader/program/src/instruction/transaction_step.rs]: done - right now `BlockParams` is separate from `Cache` so we won't have such a problem in the future.

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11558950504/artifacts/2113646925)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   61653 |      0     |      0.074 |    0.00123811  |        0      |
| getSlot                           |   53108 |      0     |      0.047 |    0.000627664 |        0      |
| getBlocks                         |   23737 |      0     |      0.072 |    0.00054514  |        0      |
| getMultipleAccounts               |   16511 |      0     |      0.07  |    0.00145279  |        0.001  |
| getBlock                          |   15225 |      0     |      0.049 |    0.00120447  |        0.001  |
| getLatestBlockhash                |   11521 |      0     |      0.078 |    0.00117715  |        0      |
| getBlockTime                      |    6263 |      0     |      0.027 |    0.000400128 |        0      |
| getTransaction                    |    5791 |      0     |      0.09  |    0.00696581  |        0.002  |
| sendTransaction                   |    5429 |      0     |      0.074 |    0.00593019  |        0.002  |
| getBalance                        |    1840 |      0     |      0.051 |    0.00122446  |        0      |
| getSignatureStatuses              |    1135 |      0     |      0.029 |    0.00102203  |        0      |
| getVersion                        |     601 |      0     |      0.037 |    0.00123295  |        0      |
| isBlockhashValid                  |     549 |      0     |      0.015 |    0.000351548 |        0      |
| simulateTransaction               |     491 |      0     |      0.004 |    0.00114257  |        0.001  |
| getSignaturesForAddress           |      74 |      0     |      0.004 |    0.00124324  |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00059375  |        0.001  |
| getTokenAccountBalance            |       8 |      0     |      0.002 |    0.001125    |        0.001  |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |        0.001  |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11572206020/artifacts/2117119641)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   58919 |      0     |      0.098 |    0.00140792  |         0     |
| getSlot                           |   55124 |      0     |      0.223 |    0.000746608 |         0     |
| getBlocks                         |   25002 |      0     |      0.038 |    0.000597352 |         0     |
| getMultipleAccounts               |   16371 |      0     |      0.094 |    0.00173893  |         0.001 |
| getBlock                          |   16083 |      0     |      0.071 |    0.00135951  |         0.001 |
| getLatestBlockhash                |   11693 |      0     |      0.096 |    0.00142701  |         0     |
| getBlockTime                      |    6049 |      0     |      0.055 |    0.000512977 |         0     |
| getTransaction                    |    5837 |      0     |      0.211 |    0.00913586  |         0.003 |
| sendTransaction                   |    5474 |      0     |      0.4   |    0.00828243  |         0.003 |
| getBalance                        |    1773 |      0     |      0.051 |    0.00140327  |         0     |
| getSignatureStatuses              |    1142 |      0     |      0.045 |    0.00114623  |         0     |
| getVersion                        |     598 |      0     |      0.046 |    0.00128763  |         0     |
| isBlockhashValid                  |     553 |      0     |      0.02  |    0.000412297 |         0     |
| simulateTransaction               |     491 |      0     |      0.004 |    0.00103666  |         0.001 |
| getSignaturesForAddress           |      77 |      0     |      0.004 |    0.00158442  |         0.002 |
| requestAirdrop                    |      32 |      0     |      0.003 |    0.00075     |         0.001 |
| getTokenAccountBalance            |       5 |      0     |      0.012 |    0.0024      |         0     |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |         0.001 |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| getClusterNodes                   |       1 |      0     |      0     |    0           |         0     |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11615306065/artifacts/2129335618)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   59164 |      0     |      0.099 |    0.00130557  |         0     |
| getSlot                           |   56678 |      0     |      0.18  |    0.000635467 |         0     |
| getBlocks                         |   26806 |      0     |      0.15  |    0.000534843 |         0     |
| getBlock                          |   17066 |      0     |      0.036 |    0.00121563  |         0.001 |
| getMultipleAccounts               |   16620 |      0     |      0.28  |    0.00145529  |         0.001 |
| getLatestBlockhash                |   12032 |      0     |      0.106 |    0.00119116  |         0     |
| getBlockTime                      |    6069 |      0     |      0.039 |    0.000405339 |         0     |
| getTransaction                    |    5875 |      0     |      0.28  |    0.00716051  |         0.002 |
| sendTransaction                   |    5526 |      0     |      0.069 |    0.00538292  |         0.003 |
| getBalance                        |    1779 |      0     |      0.064 |    0.00124227  |         0     |
| getSignatureStatuses              |    1162 |      0     |      0.042 |    0.000923408 |         0     |
| getVersion                        |     609 |      0     |      0.042 |    0.00091133  |         0     |
| isBlockhashValid                  |     563 |      0     |      0.017 |    0.00037833  |         0     |
| simulateTransaction               |     491 |      0     |      0.013 |    0.000527495 |         0     |
| getSignaturesForAddress           |      72 |      0     |      0.006 |    0.00173611  |         0.001 |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00053125  |         0.001 |
| getTokenAccountBalance            |       6 |      0     |      0.001 |    0.000666667 |         0.001 |
| getHealth                         |       3 |      0     |      0.006 |    0.00233333  |         0.001 |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0.002 |      0.002 |    0.002       |         0.002 |
| getClusterNodes                   |       1 |      0     |      0     |    0           |         0     |

</details>

**Pull request merged**